regularize EPL at origin

dangilman opened the pull request:
I just implemented the property I want in the deflection angles alpha(0, 0) = 0 by replacing nan values with zero. The function can only be nan at the origin so it works

sibirrer commented on lenstronomy/LensModel/Profiles/epl.py:
```diff
@@ -183,8 +183,27 @@ class EPLMajorAxis(LensProfileBase):
+
+        alpha_real, alpha_imag = alpha.real, alpha.imag
+
+        if isinstance(alpha_real, float) or isinstance(alpha_real, int):
```
can you simplify this with this functionality in numpy? https://numpy.org/doc/stable/reference/generated/numpy.nan_to_num.html

sibirrer commented on lenstronomy/LensModel/Profiles/epl.py:
```diff
@@ -214,7 +233,9 @@ def derivatives(self, x, y, b, t, q):
 
         # return real and imaginary part
-        return alpha.real, alpha.imag
+        alpha_real, alpha_imag = self._regularize(alpha)
```
isn't this also an issue for derivatives and Hessians?

sibirrer reviewed (changes requested):
Could you simplify things with `numpy.nan_to_num`? I don't see a specific test case that checks the case of turning nan's into zero. Thanks!

dangilman commented:
I added the numpy function. Passing in (x, y) = (0, 0) creates np.nan values inside the class that get output as zeros, but I added another test function that tests the new regularize function explicitly. And it is also an issue for hessian, but hessian is computed by calling derivatives so it is taken care of

dangilman commented:
nevermind, was looking at the wrong hessian class.... let me fix it for the hessian also

dangilman commented:
not really sure what to do with infinities... I just default them to plus/minus 1000

sibirrer commented on lenstronomy/LensModel/Profiles/epl.py:
```diff
@@ -183,8 +183,18 @@ class EPLMajorAxis(LensProfileBase):
+    def _regularize(value):
+
+        out = np.nan_to_num(value, posinf=1e+3, neginf=-1e+3)
+        if isinstance(value, float) or isinstance(value, int):
```
do you need this part?

dangilman replied:
It serves only to keep the input and output data types the same. If you supply a float, without that check you get back a numpy array. I can take it out if you want

sibirrer replied:
I had guessed that the nan_to_num keeps the same in and output format. It's worth testing and if confirmed, you don't need this.

sibirrer commented on lenstronomy/LensModel/Profiles/epl.py:
```diff
@@ -183,8 +183,18 @@ class EPLMajorAxis(LensProfileBase):
+    @staticmethod
+    def _regularize(value):
+
+        out = np.nan_to_num(value, posinf=1e+3, neginf=-1e+3)
```
posinf and neginf do not need to be specified I think.

dangilman replied:
I'll take them out

sibirrer commented on test/test_LensModel/test_Profiles/test_epl.py:
```diff
@@ -100,6 +100,42 @@ def test_static(self):
+        npt.assert_almost_equal(f_x, phi_E * x/R)
+        npt.assert_almost_equal(f_y, phi_E * y/R)
+
+        value = np.nan
```
I was more thinking of putting in (0,0) and see whether you get regular behavior out

dangilman replied:
I do that on line 111

sibirrer reviewed (changes requested):
I don't think you need the array and float/int distinction with the new routine. This will simplify things quite a bit.

sibirrer commented:
Thank you!
